yt_dlp/utils.py

```diff
@@ -5569,45 +5581,55 @@ def lazy_last(iterable):
 
         yield True, prev
 
-    def apply_path(start_obj, path, test_type=False):
+    def apply_path(start_obj, path, test_type):
         objs = (start_obj,)
         has_branched = False
 
         key = None
         for last, key in lazy_last(variadic(path, (str, bytes, dict, set))):
-            if is_user_input and key == ':':
-                key = ...
+            if is_user_input and isinstance(key, str):
+                if key == ':':
+                    key = ...
+                elif ':' not in key:
+                    if int_or_none(key) is not None:
+                        key = int(key)
+                else:
+                    key = slice(*map(int_or_none, key.split(':')))
 
             if not casesense and isinstance(key, str):
                 key = key.casefold()
 
-            if key is ... or isinstance(key, (list, tuple)) or callable(key):
-                has_branched = True
-
             if __debug__ and callable(key):
                 # Verify function signature
                 inspect.signature(key).bind(None, None)
 
-            key_func = functools.partial(apply_key, key, last)
-            objs = itertools.chain.from_iterable(map(key_func, objs))
+            new_objs = []
+            for obj in objs:
+                branch, added_iterable = apply_key(key, obj, last)
+                has_branched |= branch
+                new_objs.append(added_iterable if branch else (added_iterable,))
+
+            objs = itertools.chain.from_iterable(new_objs)
 
         if test_type and not isinstance(key, (dict, list, tuple)):
             objs = map(type_test, objs)
 
-        return has_branched, objs
-
-    def _traverse_obj(obj, path, use_list=True, test_type=True):
-        has_branched, results = apply_path(obj, path, test_type)
-        results = LazyList(x for x in results if x is not None)
+        return objs, has_branched, isinstance(key, dict)
 
+    def _traverse_obj(obj, path, allow_empty, test_type):
+        results, has_branched, is_dict = apply_path(obj, path, test_type)
+        results = LazyList(item for item in results if item or isinstance(item, (bool, int, float)))
```

Comment:
```suggestion
        results = LazyList(item for item in results if item or item in (0, False))
```
Maybe we should allow `''` too? We really only want to block `{}`/`None`, right?

Reply:
Theoretically we want to block `[]`, `{}`, `None`. How about instead of whitelisting some we blacklist these instead?

`item or not isinstance(item (list, dict, NoneType))`/`item or item not in ([], {}, None)`